go/adbc/driver/snowflake/record_reader.go

```diff
@@ -389,7 +386,15 @@ func jsonDataToArrow(ctx context.Context, bldr *array.RecordBuilder, ld gosnowfl
 					return nil, err
 				}
 
-				if tz != time.UTC {
+				snowflakeType, ok := bldr.Schema().Field(i).Metadata.GetValue(MetadataKeySnowflakeType)
+				if !ok {
+					return nil, errToAdbcErr(
+						adbc.StatusInvalidData,
+						fmt.Errorf("key %s not found in metadata for field %s", MetadataKeySnowflakeType, bldr.Schema().Field(i).Name),
+					)
+				}
```

Comment:
Are we guaranteed that `SNOWFLAKE_TYPE` will *always* appear in the metadata? Should we rely on it rather than falling back to making an inference if the metadata key isn't there?

Reply:
We do guarantee it because it is added in `rowTypesToArrowSchema`, which is always called before `jsonDataToArrow`. If this method was used elsewhere in the future without specifically producing the schema with `rowTypesToArrowSchema` first then it could cause issues though. Perhaps it is better to fail quickly in that case to ensure that kind of change isn't made in the future. This does make me think about the case where the key is present but has the default empty string value. The current behavior is to fall through to the TIMESTAMP_TZ case, which I'm not sure is correct. I'll put up a revision to this.

Reply:
That makes sense

Reply:
Ok I refactored this section. I kept the `SNOWFLAKE_TYPE` key check in place so that this can fail fast if someone changes the logic in the future. The part I changed was order of the checks for the timestamp variants.

Previously it was `timestamp_ltz` -> `timestamp_ntz` -> fallthrough `timestamp_tz`.

It turns out that `timestamp_ltz` and `timestamp_ntz` can be stored in the exact same way. Using `time.Unix(sec, nsec).In(tz)` was only changing the location metadata for the `time.Time` object. It doesn't have an effect on the output of `arrow.TimestampFromTime()` so for the nanosecond case they are equivalent.

Now we check if it's a `timestamp_tz` first. If so, we perform the offset logic, etc.

Otherwise we just extract and use the epoch nanoseconds directly, which applies to `timestamp_ltz`, `timestamp_ntz`, and the default case for unset type metadata.